subsys/net/l2/wifi/wifi_mgmt.c

```diff
@@ -403,6 +443,122 @@ void wifi_mgmt_raise_disconnect_result_event(struct net_if *iface, int status)
+			return -ENOTSUP;
+		}
+
+		return wifi_mgmt_api->start_11r_roaming(dev);
+	} else if (roaming_params.is_11k_enabled) {
+		memset(&roaming_params.neighbor_rep, 0x0, sizeof(roaming_params.neighbor_rep));
+		if (wifi_mgmt_api == NULL
+		    || wifi_mgmt_api->send_11k_neighbor_request == NULL) {
+			return -ENOTSUP;
+		}
+
+		return wifi_mgmt_api->send_11k_neighbor_request(dev, NULL);
+	} else if (wifi_mgmt_api == NULL || wifi_mgmt_api->btm_query == NULL) {
+		return -ENOTSUP;
+	}
+
+	return wifi_mgmt_api->btm_query(dev, 0x10);
+}
+
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_START_ROAMING, wifi_start_roaming);
+
+static int wifi_neighbor_rep_complete(uint32_t mgmt_request, struct net_if *iface,
+				      void *data, size_t len)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+	struct wifi_scan_params params = {0};
+
+	for (int i = 0; i < roaming_params.neighbor_rep.neighbor_cnt; i++) {
+		params.band_chan[i].channel = roaming_params.neighbor_rep.neighbor_ap[i].channel;
+		if (params.band_chan[i].channel > 14) {
+			params.band_chan[i].band = WIFI_FREQ_BAND_5_GHZ;
+		} else {
+			params.band_chan[i].band = WIFI_FREQ_BAND_2_4_GHZ;
+		}
+	}
+	if (wifi_mgmt_api == NULL || wifi_mgmt_api->candidate_scan == NULL) {
+		return -ENOTSUP;
+	}
+
+	return wifi_mgmt_api->candidate_scan(dev, &params);
+}
+
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_NEIGHBOR_REP_COMPLETE,
+				  wifi_neighbor_rep_complete);
+
+void wifi_mgmt_raise_neighbor_rep_recv_event(struct net_if *iface, char *inbuf, size_t buf_len)
+{
+	const uint8_t *buf = inbuf;
+	char event[MAX_EVENT_STR_LEN] = {0};
+	char bssid[WIFI_SSID_MAX_LEN] = {0};
+	char bssid_info[WIFI_SSID_MAX_LEN]  = {0};
+	int op_class, channel, phy_type;
+	int idx = roaming_params.neighbor_rep.neighbor_cnt;
+
+	if (!buf || buf[0] == '\0') {
+		return;
+	}
+	if (sscanf(buf, "%s bssid=%s info=%s op_class=%d chan=%d phy_type=%d",
+		   event, bssid, bssid_info, &op_class, &channel, &phy_type) == 6) {
+		int i;
+		int match  = 0;
+		size_t len = 0;
+
+		for (i = 0; i < roaming_params.neighbor_rep.neighbor_cnt; i++) {
+			if (strncmp((const char *)roaming_params.neighbor_rep.neighbor_ap[i].bssid,
+				    bssid, sizeof(bssid)) == 0) {
+				match = 1;
+				break;
+			}
+
+			if (roaming_params.neighbor_rep.neighbor_ap[i].channel == channel) {
+				match = 1;
+				break;
+			}
+		}
+		if (!match && (roaming_params.neighbor_rep.neighbor_cnt < MAX_NEIGHBOR_AP_LIMIT)) {
+			strncpy((char *)roaming_params.neighbor_rep.neighbor_ap[idx].bssid,
+				bssid, sizeof(roaming_params.neighbor_rep.neighbor_ap[idx].bssid));
+			len = strnlen(bssid, sizeof(bssid));
+			roaming_params.neighbor_rep.neighbor_ap[idx].bssid[len] = (uint8_t)'\0';
+
+			strncpy((char *)roaming_params.neighbor_rep.neighbor_ap[idx].bssid_info,
+				(bssid_info),
+				sizeof(roaming_params.neighbor_rep.neighbor_ap->bssid_info));
+			len = strnlen(bssid, sizeof(bssid));
```

Comment:
Should this be `strnlen(bssid_info, sizeof(bssid_info));`?
This could lead to missing `'\0'`, we should take into account the terminating null.
```
len = strnlen(bssid_info, sizeof(bssid_info) - 1);
```
Similar issue in line 535

Reply:
Updated.